<font size="8">Homework 2: AirBnB Document Database</font>

Group number: `3?`

Group members:
1. `Julia Karpienia : 20240514`
2. `Elizaveta Nosova : 20242132`
3. `Zofia Wójcik : 20240654`

The Homework 2 is comprised of two parts:
1. Data modelling (15 points).
2. Queries to database to answer the questions (25 points).

<font size="6">1. Data Modelling</font>

<font size="4">Congratulations! You’ve been hired as part of the new Data Engineering and Management team in the AirBNB Business Intelligence department. The company is restructuring due to unsatisfactory performance from the previous teams.

Before leaving, the head of the Data Modelling department highlighted several issues:

**Data Storage**: A lot of data about AirBNB listings is stored in a single document. While this approach has some advantages, it has also caused performance issues. Queries are slow, and the team didn’t apply patterns, which could improve performance by optimizing the data model. Indexes were also not used.

**Reviews Growth**: The number of reviews for AirBNB is growing rapidly. Currently, we overwrite reviews regularly, but the Business Intelligence department will benefit from storing all reviews and analyzing them over time.

**Data Errors**: There are errors in the data collection, such as duplicate data entries and incorrect timestamps for transactions. The new team will need to decide how to fix these issues.

**Your Role**: In your new role, you’ll need to consider how each database query is used, how often it is needed, and its impact on reads and writes. You should update the database schema to optimize for business use cases. Use tools like embedding, linking, indexes, and patterns to improve the data model. You may need to create new fields, documents, or collections. Be sure to document the pattern you’re applying and the reasons behind your decisions, especially when dealing with duplication and risks of outdated data.

**Key tasks include**:

1. Streamlining the data collection process.
2. Cleaning up the data and optimizing what will be returned for each use case.
3. Applying the correct patterns to speed up common queries.
4. Ensuring departments get accurate and relevant information from the database.
5. Sharing the updated data model schema with other departments.

**Good Practices**: [Check Chapter 6, Mastering MongoDB]

1. All newly created fields should have capitalized names.
2. New queries should work with the most up-to-date database version. If you make multiple changes, all queries should still work after the final updates.
3. For some queries, you may need to change the database schema.
4. When you are applying specific patterns, like polymorphic, subset, or bucket, name them accordingly. 
5. Document each major transformation using this format:
*“We applied {transformation name} because {reasoning behind it}. We expect {change/result} based on {observable measure, such as query speed, number of documents returned, index use, etc.}.”*

</font>


**Data Cleanup and Schema Adjustments:** [9 points in total]

1) Before working on the queries below, review the data and adjust the schema based on the typical use case described.

**Typical Use Case**: The most common use of the database is to show property listing information to customers. A query retrieves a listing document from the database. Currently, retrieving a listing takes too long. Decide what information should be included in a typical query and optimize the structure accordingly. For example, customers usually only need a sample of reviews, not all reviews (even though all reviews are stored). They also don’t need past transaction data. Update the document schema to fit this use case. This might involve creating new collections or documents.

**Data Cleanup**: Review the data for any errors (such as transactions that don’t belong to the listing) or unnecessary duplication, and clean it up where needed.

<font size="6">2. Uses for the database</font>

<font size="4">Different AirBNB departments require different analytics from our common database.
Below are the specific questions from various departments: 

**Standard Difficulty Questions:** [2 points per question]

2)	Once a month, we reward hosts with recognition. Select three superhosts with at least two listings that can accommodate more than four people.

3)	The company considers investing into property to rent. Which bed type is most common in listings with a waterfront and a dishwasher in New York?

4)	We're considering hiring someone to write reviews professionally. Who wrote the longest review in New York?

5)	To assess the security of different areas, what is the biggest and smallest (price-security deposit) difference per number of visitors at a property?

6)  Identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. This information support targeted advertising of different customer types. It is not expected to change much over time so we won’t look to update it, we just require current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Advanced Difficulty Questions (Consider database optimization for these queries):** [3 points per question]

7)	We are creating a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

8)	We plan to track our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

9)	For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We consider adding more metrics, although there is no clarity on what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

10)	We aim to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

11)	We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

**Database updates:** [2 points per question]

After optimizing the database, show how to complete the following updates. You can create fictional data. Ensure that previous data does not become stale:

12) Add a new property with a new host in one of the top 10 cities. The host selects the top 10 most common amenities to list.

13) Add a new review from one of our top 20 reviewers for this new property.

14) Add a new review metric called 'x_factor' with a score of 10. Show that the average score across all metrics is correctly calculated for this listing, using the previously developed query.


# Setup

1. run Jupyter 
2. run Docker
3. create MongoDB container in Command Prompt: make sure you don't have 'leftover' container from previous runs

        docker run --name mongodb -d -e MONGO_INITDB_ROOT_USERNAME=AzureDiamond -e MONGO_INITDB_ROOT_PASSWORD=hunter2 -p 27017:27017 mongo

4. run Studio3T

5. create New Connection OR reconnect to existing Connection:

    5.A	In Studio3T create Connection:
            a.	Press Connection -> New Connection
            b.	Insert credentials into URI field: mongodb://AzureDiamond:hunter2@localhost:27017
            c.	Press Test Connection
            d.	Assign a name to the Connection (top empty line) -> Save
            e.	Press Connect
            
    5.B In Studio3T reconnect to Connection:
            a. press Connect (top left corner)
            b. choose MongoDB connection  -> press Connect (MongoDB container must be runnning by this time)
            
6.	In Studio3T import the Database:
            a.	Press Import
            b.	Choose BSON mongodump archive 
            c.	Find your database path
            d.	Select All file formats and choose the database file ‘sampledata.archive’
            e.	Press Run: be patient while the database loads
            f.	Inspect the collections and documents

Done!

In [1]:
# Python Connector

! pip install pymongo
# or #!conda install -y pymongo

from datetime import datetime
from pprint import pprint
import time
from bson.objectid import ObjectId
from bson.decimal128 import Decimal128

from pymongo import MongoClient

from collections import Counter

user="AzureDiamond"
password="hunter2"
host="localhost"
port="27017"
protocol="mongodb"

client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

# Database check
db = client.sample_airbnb
print(f"Database info: {db}\n")
db.name 

Database info: Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'sample_airbnb')



'sample_airbnb'

[PyMongo documentation](https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html)

# Loading and Inspecting database


In [2]:
# Collections are inside our Database 'sample_analytics' (This has to be Olga's copy and paste error)

collection_list = db.list_collection_names()

print(f"The database contains {len(collection_list)} collections")
print(f"All collections: {collection_list[0:]}")
print(f"Collection {collection_list[0]} contains {db[collection_list[0]].count_documents({})} documents")

The database contains 1 collections
All collections: ['listingsAndReviews_new']
Collection listingsAndReviews_new contains 5555 documents


In [3]:
# Show all top-level fields in one listing
sample_doc = db.listingsAndReviews_new.find_one()
pprint(sample_doc)

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

Some info on what we have:
- '_id': is a string not ObjectId
- 'reviews': a list of dictionaries ('_id', 'comments', 'date', 'listing_id', 'reviewer_id', 'reviewer_name')
- 'transactions': dictionaries ('bucket_end_date', 'bucket_start_date', 'transaction_count', 'transactions' ('date', 'price'))

ChatGPT suggestions:

- Add indexes
- Create a transactions collection and flatten + clean transaction data.
- Embed top 3 reviews only into listingsAndReviews_new.
- Remove the full reviews and transactions fields.

# Data Cleanup

In [4]:
# define the original collection
raw_col = db.listingsAndReviews_new

### Duplicates

In [5]:
# Deduplicate reviews inside each listing
# Key assumption: A review is uniquely identified by (reviewer_id, date, listing_id)

total_duplicates = 0    # Count total removed reviews
deduped_docs = 0        # Count how many listing documents were cleaned

for doc in raw_col.find({}, {"_id": 1, "reviews": 1}):
    reviews = doc.get("reviews", [])
    seen_keys = set()           # Store (reviewer_id, date, listing_id)
    unique_reviews = []

    for r in reviews:
        key = (r.get("reviewer_id"), r.get("date"), doc["_id"])

        if key not in seen_keys:
            seen_keys.add(key)
            unique_reviews.append(r)
        else:
            total_duplicates += 1

    # Only update the document if changes are needed
    if len(unique_reviews) < len(reviews):
        deduped_docs += 1
        raw_col.update_one({"_id": doc["_id"]}, {"$set": {"reviews": unique_reviews}})

print(f"Deduplicated reviews in {deduped_docs} listings. Removed {total_duplicates} duplicates.")


Deduplicated reviews in 9 listings. Removed 11 duplicates.


In [6]:
# Look for a listing that has duplicate entries in its 'amenities' list
for doc in raw_col.find({"amenities": {"$exists": True}}, {"_id": 1, "name": 1, "amenities": 1}):
    amenities = doc.get("amenities", [])

    # Compare original length with length of set to detect duplicates
    if len(amenities) > len(set(amenities)):
        print(f"Listing ID: {doc['_id']}")
        print(f"Listing Name: {doc.get('name')}")
        print("Original Amenities List:")
        print(amenities)
        break  # Stop after finding the first example

Listing ID: 1001265
Listing Name: Ocean View Waikiki Marina w/prkg
Original Amenities List:
['TV', 'Cable TV', 'Wifi', 'Air conditioning', 'Pool', 'Kitchen', 'Free parking on premises', 'Elevator', 'Hot tub', 'Washer', 'Dryer', 'Essentials', 'Shampoo', 'Hangers', 'Hair dryer', 'Iron', 'Laptop friendly workspace', 'Self check-in', 'Lockbox', 'Hot water', 'Bed linens', 'Extra pillows and blankets', 'Ethernet connection', 'Microwave', 'Coffee maker', 'Refrigerator', 'Dishes and silverware', 'Cooking basics', 'Stove', 'BBQ grill', 'Garden or backyard', 'Well-lit path to entrance', 'Disabled parking spot', 'Step-free access', 'Wide clearance to bed', 'Step-free access']


In [7]:
# Amenities are a list of strings like ["TV", "Wifi", "TV"]
# We only need each amenity once, so we'll use a set to remove duplicates

amenities_cleaned = 0

for doc in raw_col.find({}, {"_id": 1, "amenities": 1}):
    amenities = doc.get("amenities", [])
    unique_amenities = list(set(amenities))  # Convert to set, then back to list

    # Only update if duplicates were removed
    if len(unique_amenities) < len(amenities):
        amenities_cleaned += 1
        raw_col.update_one({"_id": doc["_id"]}, {"$set": {"amenities": unique_amenities}})

print(f" Deduplicated amenities in {amenities_cleaned} listings.")

 Deduplicated amenities in 483 listings.


In [8]:
# Verifications are a list of strings like ["email", "phone"]
# They should be unique per host

verifications_cleaned = 0

for doc in raw_col.find({}, {"_id": 1, "host_verifications": 1}):
    verifications = doc.get("host_verifications", [])
    unique_verifications = list(set(verifications))

    if len(unique_verifications) < len(verifications):
        verifications_cleaned += 1
        raw_col.update_one({"_id": doc["_id"]}, {"$set": {"host_verifications": unique_verifications}})

print(f"Deduplicated host_verifications in {verifications_cleaned} listings.")

Deduplicated host_verifications in 0 listings.


### Invalid dates

In [9]:
# check for any dates that are not datetime
# We will check the reviews and transactions dates

# Check for invalid dates in reviews
invalid_dates = 0
for doc in raw_col.find({}, {"_id": 1, "reviews": 1}):
    reviews = doc.get("reviews", [])
    for r in reviews:
        date = r.get("date")
        if not isinstance(date, datetime):  # Check if the date is not a datetime object
            invalid_dates += 1
            print(f"Invalid review date found in listing {doc['_id']}: {date}")
# Check for invalid dates in transactions
invalid_dates = 0
for doc in raw_col.find({}, {"_id": 1, "transactions": 1}):
    transactions = doc.get("transactions", {}).get("transactions", [])
    for t in transactions:
        date = t.get("date")
        if not isinstance(date, datetime):  # Check if the date is not a datetime object
            invalid_dates += 1
            print(f"Invalid transaction date found in listing {doc['_id']}: {date}")

In [10]:
# Check for any transactions that have weird dates
# Define the year Airbnb was founded
cutoff_year = 2007

# Initialize counters
pre_2007_count = 0                  # Total number of transactions before 2007
total_txns = 0                      # Total number of valid transactions
listings_with_old_txns = 0         # Number of listings that include at least one old transaction

# Go through each listing document
for doc in raw_col.find({}, {"_id": 1, "transactions.transactions": 1}):
    txns = doc.get("transactions", {}).get("transactions", [])
    has_old_transaction = False    # Track if this listing has any old transaction

    for t in txns:
        date = t.get("date")
        if isinstance(date, datetime):  # Ensure the date is valid
            total_txns += 1

            # If the transaction occurred before 2007, update counters
            if date.year < cutoff_year:
                pre_2007_count += 1
                has_old_transaction = True

    # If this listing had at least one old transaction, count it
    if has_old_transaction:
        listings_with_old_txns += 1

# Print the summary
print(f"Total valid transactions: {total_txns}")
print(f"Transactions before {cutoff_year}: {pre_2007_count}")
print(f"Listings with at least one transaction before {cutoff_year}: {listings_with_old_txns}")

Total valid transactions: 311093
Transactions before 2007: 92374
Listings with at least one transaction before 2007: 5421


- we will need to decide what do we do with the transactions that happened before 2007

In [11]:
# Define the year Airbnb was founded
cutoff_year = 2007

# Initialize counters
pre_2007_review_count = 0               # Total number of reviews before 2007
total_valid_reviews = 0                # Total number of valid datetime reviews
listings_with_old_reviews = 0         # Number of listings with at least one old review

# Iterate through each listing document
for doc in raw_col.find({}, {"_id": 1, "reviews": 1}):
    reviews = doc.get("reviews", [])
    has_old_review = False            # Flag to track if this listing has an old review

    for r in reviews:
        date = r.get("date")
        if isinstance(date, datetime):   # Only consider properly formatted datetime values
            total_valid_reviews += 1

            if date.year < cutoff_year:
                pre_2007_review_count += 1
                has_old_review = True

    if has_old_review:
        listings_with_old_reviews += 1

# Print summary results
print(f"Total valid review dates: {total_valid_reviews}")
print(f"Reviews before {cutoff_year}: {pre_2007_review_count}")
print(f"Listings with at least one review before {cutoff_year}: {listings_with_old_reviews}")

Total valid review dates: 149781
Reviews before 2007: 0
Listings with at least one review before 2007: 0


### Transactions with invalid prices

In [12]:
# Initialize counters
invalid_price_count = 0             # Total number of invalid transactions
total_transactions = 0              # Total number of valid transactions examined
listings_with_invalid_prices = 0    # Listings that include at least one invalid price

# Iterate through all listings
for doc in raw_col.find({}, {"_id": 1, "transactions.transactions": 1}):
    txns = doc.get("transactions", {}).get("transactions", [])
    has_invalid = False             # Flag for invalid transaction within this listing

    for t in txns:
        price = t.get("price")
        try:
            price_float = float(price)
            total_transactions += 1

            if price_float <= 0:
                invalid_price_count += 1
                has_invalid = True
        except:
            # Skip prices that can't be converted to float (optional: log or count them)
            continue

    if has_invalid:
        listings_with_invalid_prices += 1

# Print results
print(f"Total transactions checked: {total_transactions}")
print(f"Transactions with price ≤ 0: {invalid_price_count}")
print(f"Listings with at least one invalid price: {listings_with_invalid_prices}")


Total transactions checked: 311093
Transactions with price ≤ 0: 0
Listings with at least one invalid price: 0


### Other inconsistencies check

In [13]:
# check for listings with missing price or name
missing_fields = raw_col.count_documents({"$or": [{"price": {"$exists": False}}, {"name": {"$exists": False}}]})
print(f"Listings missing 'price' or 'name': {missing_fields}")

Listings missing 'price' or 'name': 0


In [14]:
# check for listings with missing reviews or transactions as these could possibly be initially created but soon closed listings
no_reviews = raw_col.count_documents({"reviews": {"$exists": True, "$size": 0}})
no_transactions = raw_col.count_documents({"transactions.transactions": {"$exists": True, "$size": 0}})
print(f"Listings with no reviews: {no_reviews}")
print(f"Listings with no transactions: {no_transactions}")

Listings with no reviews: 1632
Listings with no transactions: 0


In [15]:
# check for listings with invalid room_type
# We will only allow "Entire home/apt", "Private room", and "Shared room"
allowed_types = ["Entire home/apt", "Private room", "Shared room"]
invalid_room_types = raw_col.count_documents({"room_type": {"$nin": allowed_types}})
print(f"Listings with invalid room_type: {invalid_room_types}")

Listings with invalid room_type: 0


#### MaximumNights (for question 6)

In [16]:
# check if there are any hardcoded values for maximum_nights

# We will count the number of listings with a maximum_nights value
counter = Counter()

for doc in raw_col.find({"maximum_nights": {"$exists": True}}, {"maximum_nights": 1}):
    counter[str(doc["maximum_nights"])] += 1

for val, count in counter.most_common(10):
    print(f"{val}: {count} listings")

1125: 3160 listings
30: 361 listings
365: 192 listings
90: 168 listings
60: 132 listings
7: 121 listings
1124: 95 listings
14: 92 listings
31: 81 listings
10: 78 listings


It seems like 1125 is indeed a hardcoded value for `maximum_nights` or a mistake. Value of 1124 also repeats and seems too large to be real. One option would be to cap the value at, for example, 365. However, this approach would still leave us with unrealisticly high values in question 6. Therefore, we will clean `maximum_nights` by setting values greater than 365 to null.

In [17]:
# Loop through raw_col and set maximum_nights to null if it is unrealistically high
updated = 0

for doc in raw_col.find({"maximum_nights": {"$exists": True}}):
    val = doc.get("maximum_nights")

    try:
        # Convert to int
        nights = int(val)

        # If the value is unrealistically high, set it to null
        if nights > 365:
            raw_col.update_one(
                {"_id": doc["_id"]},
                {"$set": {"maximum_nights": None}}
            )
            updated += 1

    except:
        # Skip any non-convertible values (e.g., None, strings like "N/A")
        continue

print(f"Updated {updated} documents where maximum_nights was set to null (over 365).")


Updated 3367 documents where maximum_nights was set to null (over 365).


Let's set both maximum_nights and minimum_nights to int

In [18]:
updated = 0

for doc in raw_col.find({}, {"_id": 1, "maximum_nights": 1, "minimum_nights": 1}):
    update_fields = {}

    # Convert maximum_nights to int if possible
    try:
        if "maximum_nights" in doc:
            max_nights = int(doc["maximum_nights"])
            update_fields["maximum_nights"] = max_nights
    except (ValueError, TypeError):
        pass

    # Convert minimum_nights to int if possible
    try:
        if "minimum_nights" in doc:
            min_nights = int(doc["minimum_nights"])
            update_fields["minimum_nights"] = min_nights
    except (ValueError, TypeError):
        pass

    # Update document if there are changes to apply
    if update_fields:
        raw_col.update_one({"_id": doc["_id"]}, {"$set": update_fields})
        updated += 1

print(f"Converted maximum_nights and minimum_nights to integers in {updated} documents.")


Converted maximum_nights and minimum_nights to integers in 5555 documents.


# Schema Adjustments

In [19]:
# migrate hosts listingsAndReviews_new INTO hosts

# Ensure host collection is clean
db.hosts.drop() # Drop the hosts collection if it exists in case you run this multiple times
hosts_col = db.hosts

# Use a set to track host_ids we've already processed
seen_hosts = set()
inserted_count = 0

# Extract host information from each listing
for doc in raw_col.find({}, {
    "_id": 0,  # Exclude listing ID
    "host_id": 1,
    "host_name": 1,
    "host_location": 1,
    "host_is_superhost": 1,
    "host_response_rate": 1,
    "host_response_time": 1,
    "host_total_listings_count": 1,
    "host_picture_url": 1,
    "host_verifications": 1
}):
    host_id = doc.get("host_id")

    # Skip hosts with missing ID or already inserted
    if not host_id or host_id in seen_hosts:
        continue

    # Create new host document
    host_doc = {
        "_id": host_id,  # Use host_id as _id for uniqueness
        "Name": doc.get("host_name"),
        "Location": doc.get("host_location"),
        "IsSuperhost": doc.get("host_is_superhost"),
        "ResponseRate": doc.get("host_response_rate"),
        "ResponseTime": doc.get("host_response_time"),
        "TotalListingsCount": doc.get("host_total_listings_count"),
        "PictureUrl": doc.get("host_picture_url"),
        "Verifications": list(set(doc.get("host_verifications", [])))  # remove duplicates
    }

    # Insert into the hosts collection
    hosts_col.insert_one(host_doc)
    seen_hosts.add(host_id)
    inserted_count += 1

print(f"Inserted {inserted_count} unique hosts into 'hosts' collection.")

Inserted 5104 unique hosts into 'hosts' collection.


In [20]:
# Verify host migration by checking:
# 1. Total count of unique hosts in the new collection
# 2. Sample host document
# 3. Compare with expected number of unique host_ids in the raw data

# Count of hosts in the new collection
new_host_count = hosts_col.count_documents({})
print(f"New 'hosts' collection contains {new_host_count} documents.")

# Get number of unique host_ids in the original collection
unique_host_ids = raw_col.distinct("host_id")
print(f"Original collection contains {len(unique_host_ids)} unique host_ids.")

# Show a sample host from the new collection
sample_host = hosts_col.find_one()
print("Sample migrated host document:")
from pprint import pprint
pprint(sample_host)

New 'hosts' collection contains 5104 documents.
Original collection contains 5104 unique host_ids.
Sample migrated host document:
{'IsSuperhost': False,
 'Location': 'Porto, Porto District, Portugal',
 'Name': 'Ana&Gonçalo',
 'PictureUrl': 'https://a0.muscache.com/im/pictures/fab79f25-2e10-4f0f-9711-663cb69dc7d8.jpg?aki_policy=profile_x_medium',
 'ResponseRate': 100,
 'ResponseTime': 'within an hour',
 'TotalListingsCount': 3,
 'Verifications': ['phone',
                   'reviews',
                   'offline_government_id',
                   'email',
                   'jumio',
                   'government_id'],
 '_id': '51399391'}


In [42]:
from bson.decimal128 import Decimal128

# Drop and re-create listings collection
db.listings.drop()
listings_col = db.listings

migrated_listings = 0

# Iterate through documents in raw_col
for doc in raw_col.find():
    try:
        # Extract and convert decimals
        price = float(doc["price"].to_decimal()) if isinstance(doc.get("price"), Decimal128) else float(doc.get("price", 0))
        bathrooms = float(doc["bathrooms"].to_decimal()) if isinstance(doc.get("bathrooms"), Decimal128) else float(doc.get("bathrooms", 0))

        # Collect all review score fields
        review_score_fields = {
            "Rating": doc.get("review_scores_rating"),
            "Cleanliness": doc.get("review_scores_cleanliness"),
            "Communication": doc.get("review_scores_communication"),
            "Checkin": doc.get("review_scores_checkin"),
            "Location": doc.get("review_scores_location"),
            "Value": doc.get("review_scores_value")
        }

        new_listing = {
            "_id": doc["_id"],
            "Name": doc.get("name"),
            "Address": doc.get("address"),
            "Images": doc.get("images"),
            "Price": price,
            "RoomType": doc.get("room_type"),
            "PropertyType": doc.get("property_type"),
            "Beds": doc.get("beds"),
            "Bedrooms": doc.get("bedrooms"),
            "Bathrooms": bathrooms,
            "Accommodates": doc.get("accommodates"),
            "Amenities": doc.get("amenities"),
            "Summary": doc.get("summary"),
            "ReviewScore": review_score_fields,
            "HostId": doc.get("host_id"),
            "BedType": doc.get("bed_type"),
            "SecurityDeposit": doc.get("security_deposit"),
            "MaximumNights": doc.get("maximum_nights"),
            "MinimumNights": doc.get("minimum_nights"),
            "ListingURL": doc.get("listing_url"),  # Added here
            "SampleReviews": [
                {
                    "ReviewerName": r.get("reviewer_name"),
                    "Date": r.get("date"),
                    "Comments": r.get("comments")
                }
                for r in doc.get("reviews", [])[:3]
            ]
        }

        listings_col.insert_one(new_listing)
        migrated_listings += 1

    except Exception as e:
        print(f"Skipping listing {doc['_id']} due to error: {e}")

print(f"Migrated {migrated_listings} listings into 'listings' collection.")

Migrated 5555 listings into 'listings' collection.


In [56]:
# Verify listings migration
print("Listings collection contains:", listings_col.count_documents({}), "documents")

# Show one listing to verify structure
sample_listing = listings_col.find_one()
from pprint import pprint
print("Sample migrated listing:")
pprint(sample_listing)

Listings collection contains: 5555 documents
Sample migrated listing:
{'Accommodates': 8,
 'Address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'Amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
               'Paid parking off premises',
               'Smoking allowed',
               'Pets allowed',
               'Buzzer/wireless intercom',
               'Heating',
               'Family/kid friendly',
               'Washer',
               'First aid kit',
               'Fire extinguisher',
               'Essentials',
            

Cities

In the Address field there is no designated nested field called city. A city is mentioned in Address.street, but as a part of a string consisting normally of three georpahic units where city itself can be either the first or the second geographic unit. We can assume Address.market as the field describing city (in real life there might be a more complex algorithm for deisgnating markets, but in our case the vast majority of listings belong to the same market as the city in their address).

In [20]:
# checking the unique values in Address.market field in the listings collection
for location_id in listings_col.find().distinct('Address.market'):
    pprint(location_id + ': ' + str(listings_col.count_documents({'Address.market': location_id})))

    ###NOTE need to be rerun! currently doesn't show "wrong" outputs because fixes have already been 
    # applied locally

'Barcelona: 633'
'Hong Kong: 619'
'Istanbul: 661'
'Kauai: 68'
'Maui: 154'
'Montreal: 649'
'New York: 608'
'Oahu: 253'
'Porto: 555'
'Rio De Janeiro: 606'
'Sydney: 610'
'The Big Island: 139'


We can see that 6 listings have the field Address.market with an empty string value, and 5 listings have non-informative values 'Other (Domestic)' and 'Other (International). We inspect the street field to define to which city they belong.


In [21]:
no_market = listings_col.find({"$or": [{"Address.market": ""  }, {"Address.market": "Other (Domestic)"}, {"Address.market": "Other (International)"}]},{"Address.market": 1, "Address.street": 1})
for n in no_market:
    pprint(n)
    ###NOTE need to be rerun! currently doesn't show "wrong" outputs because fixes have already been 
    # applied locally

Most of the listings with incorrect Address.market value clearly state the city in Address.field. For the values "Waverley, NSW, Australia", "Kapaa, HI, United States", "Junqueira, Aveiro, Portugal", "Maunaloa, HI, United States" the city information was defined from the territorial belonging information in Wikipedia. We set the values of Address.market to established city denominations.

In [22]:
# street: Waverley, NSW, Australia -> market: Sydney
update_1 = listings_col.update_one({"_id": "13363311"}, {"$set": {"Address.market": 'Sydney'}})
pprint(listings_col.find_one({"_id": "13363311"}, {"Address.market": 1, "Address.street": 1}))

# street: New York, NY, United States -> market: New York
update_2 = listings_col.update_one({"_id": "13528649"}, {"$set": {"Address.market": 'New York'}})
pprint(listings_col.find_one({"_id": "13528649"}, {"Address.market": 1, "Address.street": 1}))

# street: Rio de Janeiro, Rio de Janeiro, Brazil -> market: Rio de Janeiro
update_3 = listings_col.update_one({"_id": "13617872"}, {"$set": {"Address.market": 'Rio De Janeiro'}})
pprint(listings_col.find_one({"_id": "13617872"}, {"Address.market": 1, "Address.street": 1}))

# street: Montréal, Québec, Canada -> market: Montreal
update_4 = listings_col.update_one({"_id": "13702505"}, {"$set": {"Address.market": 'Montreal'}})
pprint(listings_col.find_one({"_id": "13702505"}, {"Address.market": 1, "Address.street": 1}))

# street: Barcelona, Catalunya, Spain -> market: Barcelona
update_5 = listings_col.update_one({"_id": "14161732"}, {"$set": {"Address.market": 'Barcelona'}})
pprint(listings_col.find_one({"_id": "14161732"}, {"Address.market": 1, "Address.street": 1}))

# street: Kapaa, HI, United States -> market: Kauai
update_6 = listings_col.update_one({"_id": "14234514"}, {"$set": {"Address.market": 'Kauai'}})
pprint(listings_col.find_one({"_id": "14234514"}, {"Address.market": 1, "Address.street": 1}))

# street: Junqueira, Aveiro, Portugal -> market: Porto
update_7 = listings_col.update_one({"_id": "25071762"}, {"$set": {"Address.market": 'Porto'}})
pprint(listings_col.find_one({"_id": "25071762"}, {"Address.market": 1, "Address.street": 1}))

# street: Barra de Guaratiba, Rio de Janeiro, Brazil -> market: Rio de Janeiro
update_8 = listings_col.update_one({"_id": "27265740"}, {"$set": {"Address.market": 'Rio De Janeiro'}})
pprint(listings_col.find_one({"_id": "27265740"}, {"Address.market": 1, "Address.street": 1}))

# street: Şile, İstanbul, Turkey -> market: Istanbul
update_9 = listings_col.update_one({"_id": "30520863"}, {"$set": {"Address.market": 'Istanbul'}})
pprint(listings_col.find_one({"_id": "30520863"}, {"Address.market": 1, "Address.street": 1}))

# street: Barra de Guaratiba, Rio de Janeiro, Brazil -> market: Rio de Janeiro
update_10 = listings_col.update_one({"_id": "31989710"}, {"$set": {"Address.market": 'Rio De Janeiro'}})
pprint(listings_col.find_one({"_id": "31989710"}, {"Address.market": 1, "Address.street": 1}))

# street: Maunaloa, HI, United States -> market: Maui
update_11 = listings_col.update_one({"_id": "22882443"}, {"$set": {"Address.market": 'Maui'}})
pprint(listings_col.find_one({"_id": "22882443"}, {"Address.market": 1, "Address.street": 1}))


{'Address': {'market': 'Sydney', 'street': 'Waverley, NSW, Australia'},
 '_id': '13363311'}
{'Address': {'market': 'New York', 'street': 'New York, NY, United States'},
 '_id': '13528649'}
{'Address': {'market': 'Rio De Janeiro',
             'street': 'Rio de Janeiro, Rio de Janeiro, Brazil'},
 '_id': '13617872'}
{'Address': {'market': 'Montreal', 'street': 'Montréal, Québec, Canada'},
 '_id': '13702505'}
{'Address': {'market': 'Barcelona', 'street': 'Barcelona, Catalunya, Spain'},
 '_id': '14161732'}
{'Address': {'market': 'Kauai', 'street': 'Kapaa, HI, United States'},
 '_id': '14234514'}
{'Address': {'market': 'Porto', 'street': 'Junqueira, Aveiro, Portugal'},
 '_id': '25071762'}
{'Address': {'market': 'Rio De Janeiro',
             'street': 'Barra de Guaratiba, Rio de Janeiro, Brazil'},
 '_id': '27265740'}
{'Address': {'market': 'Istanbul', 'street': 'Şile, İstanbul, Turkey'},
 '_id': '30520863'}
{'Address': {'market': 'Rio De Janeiro',
             'street': 'Barra de Guaratiba,

In [23]:
# checking the unique values in the field Address.market in the listing collection after update
for location_id in listings_col.find().distinct('Address.market'):
    pprint(location_id + ': ' + str(listings_col.count_documents({'Address.market': location_id})))

# making sure all listings have Address.market field, allowing comprehesive querying on cities
count = listings_col.count_documents({"Address.market": {"$exists": True}})
print(f'Number of listings containing Address.market field: {count}')

'Barcelona: 633'
'Hong Kong: 619'
'Istanbul: 661'
'Kauai: 68'
'Maui: 154'
'Montreal: 649'
'New York: 608'
'Oahu: 253'
'Porto: 555'
'Rio De Janeiro: 606'
'Sydney: 610'
'The Big Island: 139'
Number of listings containing Address.market field: 5555


In [23]:
# migrate all reviews into new reviews collection

# Drop and re-create reviews collection 
db.reviews.drop() # Drop the reviews collection if it exists in case you run this multiple times
reviews_col = db.reviews 
migrated_reviews = 0

# Extract all reviews from original listings
for doc in raw_col.find({}, {"_id": 1, "reviews": 1}):
    for r in doc.get("reviews", []):
        review_doc = {
            "ListingId": doc["_id"],
            "ReviewerId": r.get("reviewer_id"),
            "ReviewerName": r.get("reviewer_name"),
            "Date": r.get("date"),
            "Comments": r.get("comments")
        }
        reviews_col.insert_one(review_doc)
        migrated_reviews += 1

print(f"Migrated {migrated_reviews} reviews into 'reviews' collection.")


Migrated 149781 reviews into 'reviews' collection.


In [24]:
# Check total reviews
print("Reviews collection contains:", reviews_col.count_documents({}), "documents")

# Show one sample review
sample_review = reviews_col.find_one()
print("Sample migrated review:")
pprint(sample_review)

Reviews collection contains: 149781 documents
Sample migrated review:
{'Comments': 'A casa da Ana e do Gonçalo foram o local escolhido para a '
             'passagem de ano com um grupo de amigos. Fomos super bem '
             'recebidos com uma grande simpatia e predisposição a ajudar com '
             'qualquer coisa que fosse necessário.\r\n'
             'A casa era ainda melhor do que parecia nas fotos, totalmente '
             'equipada, com mantas, aquecedor e tudo o que pudessemos '
             'precisar.\r\n'
             'A localização não podia ser melhor! Não há melhor do que acordar '
             'de manhã e ao virar da esquina estar a ribeira do Porto.',
 'Date': datetime.datetime(2016, 1, 3, 5, 0),
 'ListingId': '10006546',
 'ReviewerId': '51483096',
 'ReviewerName': 'Cátia',
 '_id': ObjectId('67f26dd22f13084c2d7db0a7')}


In [25]:
# migrate transactions into the new transactions collection

# Drop and re-create transactions collection
db.transactions.drop() # Drop the transactions collection if it exists in case you run this multiple times
transactions_col = db.transactions
migrated_transactions = 0

# Extract transactions per listing
for doc in raw_col.find({}, {"_id": 1, "transactions.transactions": 1}):
    txns = doc.get("transactions", {}).get("transactions", [])
    for t in txns:
        try:
            txn_doc = {
                "ListingId": doc["_id"],
                "Date": t.get("date"),
                "Price": float(t.get("price"))
            }
            transactions_col.insert_one(txn_doc)
            migrated_transactions += 1
        except:
            continue  # skip invalid prices
print(f"Migrated {migrated_transactions} transactions into 'transactions' collection.")

Migrated 311093 transactions into 'transactions' collection.


In [26]:
# Check total transactions
print("Transactions collection contains:", transactions_col.count_documents({}), "documents")

# Show one sample transaction
sample_transaction = transactions_col.find_one()
print("Sample migrated transaction:")
pprint(sample_transaction)

Transactions collection contains: 311093 documents
Sample migrated transaction:
{'Date': datetime.datetime(2008, 8, 12, 0, 0),
 'ListingId': '10006546',
 'Price': 132.10637816842913,
 '_id': ObjectId('67f26f1a2f13084c2d7ff9bc')}


### Question 1

### Question 2

In [27]:
# getting listings that can accommodate more than four people
query_1 = {"$match": {"Accommodates": {"$gt": 4}}}

# joining with host collection on hostId/_id
query_2 = {
        "$lookup": {
            "from": "hosts",
            "localField": "HostId",
            "foreignField": "_id",
            "as": "host"
        }
    }

# unwinding to access host fields
query_3 = {"$unwind": "$host"}

# filtering out only superhosts
query_4 = {"$match": {"host.IsSuperhost": True}}

# grouping by host to count listings
query_5 = {
        "$group": {
            "_id": "$HostId",
            "host_name": {"$first": "$host.Name"},
            "listing_count": { "$count": {} }
        }
    }

# filtering hosts with at least 2 listings
query_6 = {"$match": {"listing_count": {"$gte": 2}}}

# optional: sorting in descending order: assume number of qualifying listing as selection criteria for hosts to be rewarede
query_7 = {"$sort": {"listing_count" : -1}}

# limmiting to 3 superhosts
query_8 = {"$limit": 3}


pipeline = [query_1, query_2, query_3, query_4, query_5, query_6, query_7, query_8]

result = listings_col.aggregate(pipeline)

for r in result:
    pprint(r)

{'_id': '36133410', 'host_name': 'Patty And Beckett', 'listing_count': 3}
{'_id': '42618840', 'host_name': 'Great Vacation Retreats', 'listing_count': 3}
{'_id': '10496350', 'host_name': 'Elite', 'listing_count': 3}


We only have two hosts satisfying the requirements. What can be the reason? Let's run the query without the match stage for superhosts and see if we can get more hosts (to illustrate, we'll set limit to 10)

In [28]:
# removing query_4 (matching superhosts) and changing limit to 10
pipeline = [query_1, query_2, query_3, query_5, query_6, query_7, {"$limit": 10}]

result = listings_col.aggregate(pipeline)

for r in result:
    pprint(r)

{'_id': '4459553', 'host_name': 'Eva&Jacques', 'listing_count': 7}
{'_id': '12243051', 'host_name': 'Sonder', 'listing_count': 5}
{'_id': '69552498', 'host_name': 'Captain Cook Resorts', 'listing_count': 4}
{'_id': '3953109', 'host_name': 'Feels Like Home', 'listing_count': 4}
{'_id': '1118742', 'host_name': 'Carol', 'listing_count': 3}
{'_id': '36133410', 'host_name': 'Patty And Beckett', 'listing_count': 3}
{'_id': '12730028', 'host_name': 'Host Wise', 'listing_count': 3}
{'_id': '961586', 'host_name': 'Mano', 'listing_count': 3}
{'_id': '179663253', 'host_name': 'Chrystel', 'listing_count': 3}
{'_id': '10496350', 'host_name': 'Elite', 'listing_count': 3}


We indeed see more hosts with 2 and more listings that can accomodate more than 4 people, but only to of them satisfy the superhost requirement. So this month only Great Vacation Retreats and Patty And Beckett will be rewarded with recognition, but the query still will correspond to the requirements if in the following months more superhosts will have 2 or more listings for more than 4 people.

**NOTE** in context of this task working on indexes on "Accommodates" for listings and partial index on "IsSuperhost" : True for hosts. Currently execution time of individual query (corresponding stage) on the second index is higher than without it

### Question 3

In [ ]:
# getting the listings which have both waterfront and a dishwasher in New York
query_1 = {"$match" : {"$and": [
        {"Amenities": {"$all": ["Waterfront", "Dishwasher"]}},  
        {"Address.market": "New York"} 
    ]
} }

#grouping by bed type
query_2 = {"$group" : {"_id": "$BedType", "sum": { "$count" : { } }}}

# sorting in descending order
query_3 = {"$sort": {"sum" : -1}}

# limiting to 1 most common type
query_4 = {"$limit": 1}

pipeline = [query_1, query_2, query_3, query_4]

result = listings_col.aggregate(pipeline)

for r in result:
    pprint(r)

{'_id': 'Real Bed', 'sum': 1}


### Question 4

In [66]:
# Match only reviews in New York with non-empty comments
query1 = {
    "$match": {
        "address.market": "New York",
        "reviews": {"$exists": True, "$ne": []}
    }
}

# Unwind reviews array
query2 = {
    "$unwind": "$reviews"
}

# Filter only reviews with non-empty comment strings
query3 = {
    "$match": {
        "reviews.comments": {"$type": "string", "$ne": ""}
    }
}

# Add field for review length
query4 = {
    "$addFields": {
        "review_length": {"$strLenCP": "$reviews.comments"}
    }
}

# Sort by review length descending
query5 = {
    "$sort": {"review_length": -1}
}

# Limit to top 1 longest comment 
query6 = {
    "$limit": 1
}

# Project relevant fields
query7 = {
    "$project": {
        "_id": 0,
        "Reviewer": "$reviews.reviewer_name",
        "Review": "$reviews.comments",
        "Length": "$review_length"
    }
}

pipeline = [query1, query2, query3, query4, query5, query6, query7]

# Execute and pretty-print result
result = list(db.listingsAndReviews_new.aggregate(pipeline))
pprint(result)


[{'Length': 4665,
  'Review': 'This alleged “full bedroom”; Private room with the red couch '
            'shown on the listing’s pictures (which the couch is not in the '
            'room, at least it wasn’t when I arrived) is a complete no go for '
            'me. 5 stars DOWN. This was a booking that I made less than 48 '
            'hours from my scheduled arrival. I have been using this AIRBNB '
            'service now for 3 years, so I would like to think I’m pretty '
            'familiar with how things work. The raised eyebrows began when '
            'Carlos sent me a text message outside of AirBnB directly from a '
            'cell phone number to my cell phone number the day of check in '
            'requesting that I send him a picture of my US ID or Passport, as '
            'it is needed for self-check in. Again, since I have reserved with '
            'several hosts in the past who only did self-check in and never '
            'was asked this, I immediately as

### Question 5

In [ ]:
# Only keep docs with required fields
query1 = {
    "$match": {
        "price": {"$exists": True, "$ne": None},
        "security_deposit": {"$exists": True, "$ne": None},
        "accommodates": {"$exists": True, "$ne": None},
        "address.market": {"$exists": True}
    }
}

# Project the fields we need early
query2 = {
    "$project": {
        "_id": 1,
        "market": "$address.market",
        "accommodates": 1,
        "price": 1,
        "security_deposit": 1,
        "diff": {"$subtract": ["$price", "$security_deposit"]}
    }
}

# Sort by accommodates + diff descending to get max per group
query3 = {
    "$sort": {
        "accommodates": 1,
        "diff": -1
    }
}

# Get the max diff per accommodates
query4 = {
    "$group": {
        "_id": "$accommodates",
        "max_diff": {"$first": "$diff"},
        "max_listing_id": {"$first": "$_id"},
        "max_market": {"$first": "$market"}
    }
}

# Sort by accommodates + diff ascending to get min per group
query5 = {
    "$sort": {
        "accommodates": 1,
        "diff": 1
    }
}

# Get the min diff per accommodates
query6 = {
    "$group": {
        "_id": "$accommodates",
        "min_diff": {"$first": "$diff"},
        "min_listing_id": {"$first": "$_id"},
        "min_market": {"$first": "$market"}
    }
}

# Merge both pipelines via $facet to compute max and min separately, reusing the projection and avoiding conflicts
query7 = {
    "$facet": {
        "max_diffs": [
            query3, query4
        ],
        "min_diffs": [
            query5, query6
        ]
    }
}

pipeline = [query1, query2, query7]

result = list(raw_col.aggregate(pipeline))
pprint(result)

[{'max_diffs': [{'_id': 5,
                 'max_diff': Decimal128('8501.00'),
                 'max_listing_id': '22200454',
                 'max_market': 'Hong Kong'},
                {'_id': 3,
                 'max_diff': Decimal128('1248.00'),
                 'max_listing_id': '21974572',
                 'max_market': 'Hong Kong'},
                {'_id': 15,
                 'max_diff': Decimal128('227.00'),
                 'max_listing_id': '20701559',
                 'max_market': 'Istanbul'},
                {'_id': 16,
                 'max_diff': Decimal128('2974.00'),
                 'max_listing_id': '23219274',
                 'max_market': 'Istanbul'},
                {'_id': 9,
                 'max_diff': Decimal128('1350.00'),
                 'max_listing_id': '13603801',
                 'max_market': 'Hong Kong'},
                {'_id': 2,
                 'max_diff': Decimal128('2661.00'),
                 'max_listing_id': '20940124',
                 'ma

### Question 6

In [30]:
# Check the number of unique cities to see if we can list average duration per city 
unique_cities = listings_col.distinct("Address.market")
print(f"Number of unique cities: {len(unique_cities)}")

Number of unique cities: 15


Since there are just 15 unique cities it is viable to list "average duration of stay" (or rather the average MaximumNights) per type of property per city:

In [31]:
# First, we will filter listings with valid fields and non-null MaximumNights
query_1 = {
    "$match": {
        "MaximumNights": {"$exists": True, "$ne": None},
        "PropertyType": {"$exists": True, "$ne": None},
        "Address.market": {"$exists": True, "$ne": None}
    }
}

# Group by PropertyType and City, compute average MaximumNights
query_2 = {
    "$group": {
        "_id": {
            "PropertyType": "$PropertyType",
            "City": "$Address.market"
        },
        "avg_max_nights": {"$avg": "$MaximumNights"}
    }
}

# Sort by PropertyType and then by average MaximumNights in descending order
query_3 = {
    "$sort": {
        "_id.PropertyType": 1,
        "avg_max_nights": -1
    }
}

pipeline_avg_by_city = [query_1, query_2, query_3]

results_by_city = listings_col.aggregate(pipeline_avg_by_city)

for r in results_by_city:
    pprint(r)

{'_id': {'City': 'Rio De Janeiro', 'PropertyType': 'Aparthotel'},
 'avg_max_nights': 212.5}
{'_id': {'City': 'Istanbul', 'PropertyType': 'Aparthotel'},
 'avg_max_nights': 164.0}
{'_id': {'City': 'Barcelona', 'PropertyType': 'Aparthotel'},
 'avg_max_nights': 90.0}
{'_id': {'City': 'Oahu', 'PropertyType': 'Aparthotel'}, 'avg_max_nights': 60.0}
{'_id': {'City': 'Kauai', 'PropertyType': 'Apartment'},
 'avg_max_nights': 151.33333333333334}
{'_id': {'City': 'Oahu', 'PropertyType': 'Apartment'}, 'avg_max_nights': 122.12}
{'_id': {'City': 'Barcelona', 'PropertyType': 'Apartment'},
 'avg_max_nights': 117.63546798029557}
{'_id': {'City': 'Hong Kong', 'PropertyType': 'Apartment'},
 'avg_max_nights': 112.97744360902256}
{'_id': {'City': 'The Big Island', 'PropertyType': 'Apartment'},
 'avg_max_nights': 88.75}
{'_id': {'City': 'Istanbul', 'PropertyType': 'Apartment'},
 'avg_max_nights': 83.29113924050633}
{'_id': {'City': 'Maui', 'PropertyType': 'Apartment'},
 'avg_max_nights': 80.21428571428571}
{

Now, let's answer the main question - lowest and lowest MaximumNights per property type:

In [32]:
# Select listings with valid fields:
query_1 = {
    "$match": {
        "MaximumNights": {"$exists": True, "$ne": None},
        "PropertyType": {"$exists": True, "$ne": None},
        "Address.market": {"$exists": True, "$ne": None}
    }
}

# Group by (PropertyType, City) and calculate the average stay
query_2 = {
    "$group": {
        "_id": {
            "PropertyType": "$PropertyType",
            "City": "$Address.market"
        },
        "avg_max_nights": {"$avg": "$MaximumNights"}
    }
}

# Sort the results by PropertyType (to group later) in descending order of average stay
query_3 = {
    "$sort": {
        "_id.PropertyType": 1,
        "avg_max_nights": -1
    }
}

# Group again by PropertyType only.
query_4 = {
    "$group": {
        "_id": "$_id.PropertyType",
        "max_city": {
            "$first": {
                "City": "$_id.City",
                "AverageDuration": "$avg_max_nights"
            }
        },
        "min_city": {
            "$last": {
                "City": "$_id.City",
                "AverageDuration": "$avg_max_nights"
            }
        }
    }
}

pipeline_min_max = [query_1, query_2, query_3, query_4]

results_min_max = listings_col.aggregate(pipeline_min_max)

for r in results_min_max:
    pprint(r)

{'_id': 'Bungalow',
 'max_city': {'AverageDuration': 165.0, 'City': 'Oahu'},
 'min_city': {'AverageDuration': 29.5, 'City': 'Sydney'}}
{'_id': 'Condominium',
 'max_city': {'AverageDuration': 157.41666666666666, 'City': 'Hong Kong'},
 'min_city': {'AverageDuration': 30.0, 'City': ''}}
{'_id': 'Chalet',
 'max_city': {'AverageDuration': 30.0, 'City': 'Istanbul'},
 'min_city': {'AverageDuration': 30.0, 'City': 'Istanbul'}}
{'_id': 'Camper/RV',
 'max_city': {'AverageDuration': 30.0, 'City': 'Porto'},
 'min_city': {'AverageDuration': 30.0, 'City': 'Porto'}}
{'_id': 'Nature lodge',
 'max_city': {'AverageDuration': 90.0, 'City': 'Istanbul'},
 'min_city': {'AverageDuration': 90.0, 'City': 'Istanbul'}}
{'_id': 'Apartment',
 'max_city': {'AverageDuration': 151.33333333333334, 'City': 'Kauai'},
 'min_city': {'AverageDuration': 12.5, 'City': ''}}
{'_id': 'Hotel',
 'max_city': {'AverageDuration': 52.0, 'City': 'Hong Kong'},
 'min_city': {'AverageDuration': 35.0, 'City': 'Other (International)'}}
{'_

The average `MaximumNights` still seem too large to be real, however, having so little context does not make it reasonable to make more changes to the variable than we already did. Maybe we should have capped the `MaximumNights` lower than 365? Using `MinimumNights` to calculate the averages gave very low values on the other hand. Therefore, this is the best we could do.

### Question 7

### Question 8

### Question 9

### Question 10

In [33]:
# Define the target listing ID and date range
listing_id = "10009999" 
start_date = datetime(2008, 1, 1)
end_date = datetime(2013, 1, 1)

# Match the listing
query_1 = {
    "$match": {
        "_id": listing_id
    }
}

# Unwind the embedded transactions array
query_2 = {
    "$unwind": "$transactions.transactions"
}

# Filter by date range
query_3 = {
    "$match": {
        "transactions.transactions.date": {
            "$gte": start_date,
            "$lte": end_date
        }
    }
}

# Group and compute average and count
query_4 = {
    "$group": {
        "_id": "$_id",  # Listing ID
        "AverageTransactionValue": {
            "$avg": {
                "$toDouble": "$transactions.transactions.price"
            }
        },
        "TransactionCount": {"$sum": 1}
    }
}

# Final pipeline
pipeline = [query_1, query_2, query_3, query_4]

# Explain the execution plan for the pipeline
plan = db.command({
    "explain": {
        "aggregate": "listingsAndReviews_new",  # raw_col collection name
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})

print("Execution Plan:")
pprint(plan)

Execution Plan:
{'command': {'$db': 'sample_airbnb',
             'aggregate': 'listingsAndReviews_new',
             'cursor': {},
             'pipeline': [{'$match': {'_id': '10009999'}},
                          {'$unwind': '$transactions.transactions'},
                          {'$match': {'transactions.transactions.date': {'$gte': datetime.datetime(2008, 1, 1, 0, 0),
                                                                         '$lte': datetime.datetime(2013, 1, 1, 0, 0)}}},
                          {'$group': {'AverageTransactionValue': {'$avg': {'$toDouble': '$transactions.transactions.price'}},
                                      'TransactionCount': {'$sum': 1},
                                      '_id': '$_id'}}]},
 'explainVersion': '1',
 'ok': 1.0,
 'queryShapeHash': '47A6329503D1E6AC7EB9A15DB00D4C549DC57326D5BC654BBBDE98511B531BF2',
 'serverInfo': {'gitVersion': 'cb9e2e5e552ee39dea1e39d7859336456d0c9820',
                'host': '8b34464cc321',
          

In [34]:
# Define the target listing ID and date range
listing_id = "10009999"
start_date = datetime(2008, 1, 1)
end_date = datetime(2013, 1, 1)

# Step 1 – Match the listing and date range
query_1 = {
    "$match": {
        "ListingId": listing_id,
        "Date": {"$gte": start_date, "$lte": end_date}
    }
}

# Step 2 – Group and calculate average and count
query_2 = {
    "$group": {
        "_id": "$ListingId",
        "AverageTransactionValue": {"$avg": "$Price"},
        "TransactionCount": {"$sum": 1}
    }
}

# Final pipeline
pipeline = [query_1, query_2]

# EXPLAIN the execution plan for this pipeline
plan = db.command({
    "explain": {
        "aggregate": "transactions",  # Name of the transactions collection
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})

print("Execution Plan:")
pprint(plan)

Execution Plan:
{'command': {'$db': 'sample_airbnb',
             'aggregate': 'transactions',
             'cursor': {},
             'pipeline': [{'$match': {'Date': {'$gte': datetime.datetime(2008, 1, 1, 0, 0),
                                               '$lte': datetime.datetime(2013, 1, 1, 0, 0)},
                                      'ListingId': '10009999'}},
                          {'$group': {'AverageTransactionValue': {'$avg': '$Price'},
                                      'TransactionCount': {'$sum': 1},
                                      '_id': '$ListingId'}}]},
 'executionStats': {'executionStages': {'closes': 1,
                                        'executionTimeMillisEstimate': 309,
                                        'inputStage': {'closes': 1,
                                                       'executionTimeMillisEstimate': 309,
                                                       'inputStage': {'closes': 1,
                                      

So far, the transactions collection is much slower than the original one. It is because MongoDB has to scan the entire collection due to the lack of indexes. To solve that, we will add indexes in the transactions collection

In [35]:
transactions_col.create_index([("ListingId", 1), ("Date", 1)])

'ListingId_1_Date_1'

In [ ]:
# Define the target listing ID and date range
listing_id = "10009999"
start_date = datetime(2008, 1, 1)
end_date = datetime(2013, 1, 1)

# Match the listing and date range
query_1 = {
    "$match": {
        "ListingId": listing_id,
        "Date": {"$gte": start_date, "$lte": end_date}
    }
}

# Group and calculate average and count
query_2 = {
    "$group": {
        "_id": "$ListingId",
        "AverageTransactionValue": {"$avg": "$Price"},
        "TransactionCount": {"$sum": 1}
    }
}

pipeline = [query_1, query_2]

# explain the execution plan for this pipeline
plan = db.command({
    "explain": {
        "aggregate": "transactions",  # Name of the transactions collection
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})

print("Execution Plan:")
pprint(plan)

Execution Plan:
{'command': {'$db': 'sample_airbnb',
             'aggregate': 'transactions',
             'cursor': {},
             'pipeline': [{'$match': {'Date': {'$gte': datetime.datetime(2008, 1, 1, 0, 0),
                                               '$lte': datetime.datetime(2013, 1, 1, 0, 0)},
                                      'ListingId': '10009999'}},
                          {'$group': {'AverageTransactionValue': {'$avg': '$Price'},
                                      'TransactionCount': {'$sum': 1},
                                      '_id': '$ListingId'}}]},
 'executionStats': {'executionStages': {'closes': 1,
                                        'executionTimeMillisEstimate': 7,
                                        'inputStage': {'closes': 1,
                                                       'executionTimeMillisEstimate': 7,
                                                       'inputStage': {'closes': 1,
                                          

Even though we created separate collection for transactions and added indexes, the search on original collection is still faster. It is because in the original collection we are retrieving only one document from disk and the transaction array is already inside that document, so it’s all in memory instantly. While in transactions_col, MongoDB has to scan for indexes and only then retrieve the documents. Therefore, we will use the original collection for our search engine. However, if we wished to develop a search engine, which retrieves transactional information about more than one listing or if the listings had a huge number of transactions, transactions collection would probably outperform listingsAndReviews_new collection.

In [37]:
# Define the target listing ID and date range
target_listing_id = "10009999" 
start_date = datetime(2008, 1, 1)
end_date = datetime(2013, 1, 1)

# Match the listing
query_1 = {
    "$match": {
        "_id": target_listing_id
    }
}

# Unwind the embedded transactions array
query_2 = {
    "$unwind": "$transactions.transactions"
}

# Filter by date range
query_3 = {
    "$match": {
        "transactions.transactions.date": {
            "$gte": start_date,
            "$lte": end_date
        }
    }
}

# Group and compute average and count
query_4 = {
    "$group": {
        "_id": "$_id",  # Listing ID
        "AverageTransactionValue": {
            "$avg": {
                "$toDouble": "$transactions.transactions.price"
            }
        },
        "TransactionCount": {"$sum": 1}
    }
}

pipeline = [query_1, query_2, query_3, query_4]

result = raw_col.aggregate(pipeline)
for doc in result:
    pprint(doc)

{'AverageTransactionValue': 33.86048272085974,
 'TransactionCount': 22,
 '_id': '10009999'}


### Question 11

Some notes about the question interpretation:
- "top 10 cities": They will be determined based on the number of listings per city. We want to display the cities that have "the most to offer" based on supply.
- "top 3 properties": We will display the properties based on the sum of all review scores. Most likely we will get some ties. One option to determine the ties would be to always choose the top 3 to display randomly. However, since our priority is to keep the search fast, we will always show break ties by `_id`. This will inavetibly promote some listings more than the others, but it is also likely that after more guests leave their reviews in those "more promoted" properties, they will eventually drop down. The bottom line is, we want to keep the search criteria as simple as possible to achieve the most efficient performance we can.
- summary for the city: number of listings and average price per property type
- basic information about the top 3 properties: name, price, property type, review scores rating (just one type of score to keep it simple), listing URL

In [57]:
# Filter listings with required fields
query1 = {
    "$match": {
        "address.market": {"$exists": True, "$ne": None},
        "review_scores_rating": {"$exists": True}
    }
}

# Count listings per city
query2 = {
    "$group": {
        "_id": "$address.market",
        "ListingCount": {"$sum": 1}
    }
}

# Sort cities by count in descending order
query3 = {
    "$sort": {"ListingCount": -1}
}

# Limit to top 10 cities
query4 = {
    "$limit": 10
}

# Lookup listings in each city
query5 = {
    "$lookup": {
        "from": "listingsAndReviews_new",
        "localField": "_id",
        "foreignField": "address.market",
        "as": "city_listings"
    }
}

# Unwind listings
query6 = {
    "$unwind": "$city_listings"
}

# Compute basic city summary info
query7 = {
    "$group": {
        "_id": {
            "City": "$_id",
            "PropertyType": "$city_listings.property_type"
        },
        "AvgPrice": {"$avg": "$city_listings.price"},
        "TypeCount": {"$sum": 1},
        "Listings": {"$push": "$city_listings"}
    }
}

# Group listings and summary per city
query8 = {
    "$group": {
        "_id": "$_id.City",
        "Summary": {
            "$push": {
                "PropertyType": "$_id.PropertyType",
                "AvgPrice": "$AvgPrice",
                "Count": "$TypeCount"
            }
        },
        "Listings": {"$push": "$Listings"}
    }
}

# Flatten the nested listing array
query9 = {
    "$unwind": "$Listings"
}
query10 = {
    "$unwind": "$Listings"
}

# Project relevant fields for scoring
query11 = {
    "$project": {
        "City": "$_id",
        "Summary": 1,
        "ListingId": "$Listings._id",
        "Name": "$Listings.name",
        "Price": "$Listings.price",
        "PropertyType": "$Listings.property_type",
        "Rating": "$Listings.review_scores_rating",
        "ListingURL": "$Listings.listing_url",
        "ScoreSum": {
            "$add": [
                "$Listings.review_scores_rating",
                "$Listings.review_scores_cleanliness",
                "$Listings.review_scores_communication",
                "$Listings.review_scores_checkin",
                "$Listings.review_scores_location",
                "$Listings.review_scores_value"
            ]
        }
    }
}

# Sort by City, then ScoreSum, then _id
query12 = {
    "$sort": {
        "City": 1,
        "ScoreSum": -1,
        "ListingId": 1
    }
}

# Group and keep top 3 per city
query13 = {
    "$group": {
        "_id": "$City",
        "Summary": {"$first": "$Summary"},
        "TopListings": {"$push": {
            "ListingId": "$ListingId",
            "Name": "$Name",
            "Price": "$Price",
            "PropertyType": "$PropertyType",
            "Rating": "$Rating",
            "ListingURL": "$ListingURL"
        }}
    }
}

# Slice top 3 listings
query14 = {
    "$project": {
        "_id": 1,
        "Summary": 1,
        "TopListings": {"$slice": ["$TopListings", 3]}
    }
}

pipeline = [
    query1, query2, query3, query4,
    query5, query6, query7, query8,
    query9, query10, query11, query12,
    query13, query14
]

plan = db.command({
    "explain": {
        "aggregate": "listingsAndReviews_new",
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})
print("Execution Plan for raw_col:")
pprint(plan)

OperationFailure: $push used too much memory and cannot spill to disk. Memory limit: 104857600 bytes, full error: {'ok': 0.0, 'errmsg': '$push used too much memory and cannot spill to disk. Memory limit: 104857600 bytes', 'code': 146, 'codeName': 'ExceededMemoryLimit'}

In [58]:
# now let's see the execution performance for listings_col and determine which collection to use 

# Filter listings with market and score
query1 = {
    "$match": {
        "Address.market": {"$exists": True, "$ne": None},
        "ReviewScore.Rating": {"$exists": True}
    }
}

# Count listings per city
query2 = {
    "$group": {
        "_id": "$Address.market",
        "ListingCount": {"$sum": 1}
    }
}

# Sort cities by count
query3 = {
    "$sort": {"ListingCount": -1}
}

# Limit to top 10 cities
query4 = {
    "$limit": 10
}

# Lookup listings from listings_col
query5 = {
    "$lookup": {
        "from": "listings",
        "localField": "_id",
        "foreignField": "Address.market",
        "as": "city"
    }
}

# Unwind listings
query6 = {
    "$unwind": "$city"
}

# Compute basic city summary info
query7 = {
    "$group": {
        "_id": {
            "City": "$_id",
            "PropertyType": "$city.PropertyType"
        },
        "AvgPrice": {"$avg": "$city.Price"},
        "TypeCount": {"$sum": 1},
        "Listings": {"$push": "$city"}
    }
}

# Group listings and summary per city
query8 = {
    "$group": {
        "_id": "$_id.City",
        "Summary": {
            "$push": {
                "PropertyType": "$_id.PropertyType",
                "AvgPrice": "$AvgPrice",
                "Count": "$TypeCount"
            }
        },
        "Listings": {"$push": "$Listings"}
    }
}

# Flatten the nested listing array
query9 = {
    "$unwind": "$Listings"
}
query10 = {
    "$unwind": "$Listings"
}

# Project relevant fields
query11 = {
    "$project": {
        "City": "$_id",
        "Summary": 1,
        "ListingId": "$Listings._id",
        "Name": "$Listings.Name",
        "Price": "$Listings.Price",
        "PropertyType": "$Listings.PropertyType",
        "Rating": "$Listings.ReviewScore.Rating",
        "ListingURL": "$Listings.ListingURL",
        "ScoreSum": {
            "$add": [
                "$Listings.ReviewScore.Rating",
                "$Listings.ReviewScore.Cleanliness",
                "$Listings.ReviewScore.Communication",
                "$Listings.ReviewScore.Checkin",
                "$Listings.ReviewScore.Location",
                "$Listings.ReviewScore.Value"
            ]
        }
    }
}

# Sort to break ties
query12 = {
    "$sort": {
        "City": 1,
        "ScoreSum": -1,
        "ListingId": 1
    }
}

# Group by city and push top listings
query13 = {
    "$group": {
        "_id": "$City",
        "Summary": {"$first": "$Summary"},
        "TopListings": {"$push": {
            "ListingId": "$ListingId",
            "Name": "$Name",
            "Price": "$Price",
            "PropertyType": "$PropertyType",
            "Rating": "$Rating",
            "ListingURL": "$ListingURL"
        }}
    }
}

# Slice top 3
query14 = {
    "$project": {
        "_id": 1,
        "Summary": 1,
        "TopListings": {"$slice": ["$TopListings", 3]}
    }
}

pipeline = [
    query1, query2, query3, query4,
    query5, query6, query7, query8,
    query9, query10, query11, query12,
    query13, query14
]

plan = db.command({
    "explain": {
        "aggregate": "listings",
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})
print("Execution Plan for listings_col:")
pprint(plan)

Execution Plan for listings_col:
{'command': {'$db': 'sample_airbnb',
             'aggregate': 'listings',
             'cursor': {},
             'pipeline': [{'$match': {'Address.market': {'$exists': True,
                                                         '$ne': None},
                                      'ReviewScore.Rating': {'$exists': True}}},
                          {'$group': {'ListingCount': {'$sum': 1},
                                      '_id': '$Address.market'}},
                          {'$sort': {'ListingCount': -1}},
                          {'$limit': 10},
                          {'$lookup': {'as': 'city',
                                       'foreignField': 'Address.market',
                                       'from': 'listings',
                                       'localField': '_id'}},
                          {'$unwind': '$city'},
                          {'$group': {'AvgPrice': {'$avg': '$city.Price'},
                                   

After looking over the execution stats we have a clear winner - using listings collection did not run out of memory as compared to when using listingsAndReviews_new collection. To further optimize the search, we will add indexes to two fields - Address.market to speed up initial filtering.

In [50]:
# Index on Address.market
listings_col.create_index("Address.market")

'Address.market_1'

In [59]:
# Now let's see the execution performance for listings_col after adding indexes
plan = db.command({
    "explain": {
        "aggregate": "listings",
        "pipeline": pipeline,
        "cursor": {}
    },
    "verbosity": "executionStats"
})
print("Execution Plan for listings_col:")
pprint(plan)

Execution Plan for listings_col:
{'command': {'$db': 'sample_airbnb',
             'aggregate': 'listings',
             'cursor': {},
             'pipeline': [{'$match': {'Address.market': {'$exists': True,
                                                         '$ne': None},
                                      'ReviewScore.Rating': {'$exists': True}}},
                          {'$group': {'ListingCount': {'$sum': 1},
                                      '_id': '$Address.market'}},
                          {'$sort': {'ListingCount': -1}},
                          {'$limit': 10},
                          {'$lookup': {'as': 'city',
                                       'foreignField': 'Address.market',
                                       'from': 'listings',
                                       'localField': '_id'}},
                          {'$unwind': '$city'},
                          {'$group': {'AvgPrice': {'$avg': '$city.Price'},
                                   

RERUN AGAIN AND CHANGE IF NEEDED: Even though indexing didn’t speed things up much with our current dataset, we’re keeping it since the DB will grow. With more docs, the index will likely make a real difference and avoid expensive collection scans later. Now, let's eventually run the pipeline and see the results in a pretty format

In [60]:
# Run the pipeline
results = db.listings.aggregate(pipeline)

# Display results nicely
for city in results:
    print(f"\nCity: {city['_id']}")
    print("Summary by Property Type:")
    for entry in city["Summary"]:
        print(f"  - {entry['PropertyType']}: {entry['Count']} listings, Avg. Price: ${entry['AvgPrice']:.2f}")
    
    print("\nTop 3 Listings:")
    for listing in city["TopListings"]:
        print(f"  • {listing['Name']} | {listing['PropertyType']} | ${listing['Price']} | Rating: {listing['Rating']}")
        print(f"    URL: {listing['ListingURL']}")


City: Barcelona
Summary by Property Type:
  - Tiny house: 1 listings, Avg. Price: $110.00
  - Aparthotel: 3 listings, Avg. Price: $189.33
  - Loft: 19 listings, Avg. Price: $89.47
  - House: 8 listings, Avg. Price: $84.62
  - Farm stay: 1 listings, Avg. Price: $100.00
  - Boat: 1 listings, Avg. Price: $35.00
  - Casa particular (Cuba): 3 listings, Avg. Price: $1051.00
  - Townhouse: 1 listings, Avg. Price: $135.00
  - Villa: 1 listings, Avg. Price: $95.00
  - Serviced apartment: 22 listings, Avg. Price: $211.86
  - Guesthouse: 2 listings, Avg. Price: $33.00
  - Other: 3 listings, Avg. Price: $31.67
  - Boutique hotel: 3 listings, Avg. Price: $109.33
  - Apartment: 534 listings, Avg. Price: $91.85
  - Bed and breakfast: 8 listings, Avg. Price: $37.75
  - Condominium: 15 listings, Avg. Price: $140.73
  - Guest suite: 4 listings, Avg. Price: $83.00
  - Hostel: 3 listings, Avg. Price: $95.00

Top 3 Listings:
  • Cozy bedroom Sagrada Familia | Apartment | $20.0 | Rating: 100
    URL: https

### Question 12

### Question 13

### Question 14